Automattic/woocommerce-payments
# Payment Requests - Add support for bookable products on product page

Fixes #4444

#### Changes proposed in this Pull Request

This aims to allow bookable products to be purchased via Payment Request buttons. Bookable products can require complex calculations. Woo Bookings makes this calculation on the server then returns a block on HTML that includes the cost of the booking. Unfortunately, there's no good way to access the cost of the booking from the response of that calculation for use when initializing the payment request button. For that reason, the best option I've found is to listen for the calculation to complete, add the product to the cart, then get costs from the cart. Subsequently, the product is removed from the cart and the booking(Woo Bookings creates a 'booking' with `in-cart` status with each add to cart) is removed. Woo Bookings uses the `in-cart` status to prevent the slot from being booked while in the cart. This normally expires after 60 minutes, but it is applicable in this temporary add to cart scenario.

The flow is:

1. Bookings cost calculation is triggered by Woo Bookings
2. The product is added to the cart
3. The product is removed from the cart and booking is removed
4. `total` and `displayItems` are used to re-init the payment request button
5. Payment request button is clicked and the normal PRB add to cart flow begins









#### Testing instructions




Woo Bookings currently [filters `wcpay_payment_request_is_product_supported`](https://github.com/woocommerce/woocommerce-bookings/blob/1995dd787591bb6ee49ba75c35a8418d52ef7b75/includes/class-wc-bookings-init.php#L41) to return false to prevent PRBs from appearing on bookable products. Either edit the callback function there or override the filter somewhere:

```
add_filter( 'wcpay_woopay_button_is_product_supported', function() {
    return true;
}, 9, 2 );
```

If this is merged, Woo Bookings can add support on that end.

* Install [Woo Bookings](https://woo.com/products/woocommerce-bookings)
* Add some different Bookable products using the templates available at **Bookings > Add Product**
* Test different configurations of totals with tax/no tax, shipping/no shipping
* Ensure the totals in the payment request sheet are correct or matching those that appear in the cart
* After the bookings cost calculation completes(before clicking the PRB) for a product, check **Bookings > In-cart** in the admin. There shouldn't be a new In-Cart booking appearing until after the PRB is clicked.

-------------------


**Post merge**
